psydac/linalg/solvers.py

```diff
@@ -1134,13 +1382,13 @@ def __init__(self, A, *, x0=None, tol=None, atol=None, btol=None, maxiter=1000,
         self._domain = domain
         self._codomain = codomain
         self._solver = 'cg'
-        self._options = {"x0":x0, "tol":tol, "atol":atol, "btol":btol, 
+        self._options = {"x0":x0, "tol":tol, "atol":atol, "btol":btol,
                          "maxiter":maxiter, "conlim":conlim, "verbose":verbose}
         self._check_options(**self._options)
         self._info = None
         self._successful = None
-        tmps_codomain = {key: A.codomain.zeros() for key in ("uh", "uc")}
-        tmps_domain = {key: A.domain.zeros() for key in ("v", "vh", "h", "hbar")}
+        tmps_codomain = {key: domain.zeros() for key in ("uh", "uc")}
+        tmps_domain = {key: codomain.zeros() for key in ("v", "vh", "h", "hbar")}
```

Comment:
Maybe the names of these dictionaries should be swapped?

Reply:
Ah yes, the names of the dics should be changed, I can do that now.
I swapped the names because they were wrong, I confused A.domain and self._domain. That this didn't cause tests to fail is probably because I never used InverseLinearOperators that map from V -> W, were V and W are mathematically the same, but different objects, or at least have the same dimension, but instead only tested the case V -> V. I shall quickly include a case in my test file that would cause without this change.

Reply:
@yguclu @tomcaruso There`s unfortunately a problem in two algorithms, the ones from `pcg` and `lsmr`, here`s the pcg issue:
```
   # The first line asserts isinstance(v, A.codomain), the second asserts r.space == v.space
   A.dot(x, out=v) # line 428
   r       -= v # line 430
   ...
   # The first two lines assert s.space == r.space, the third asserts s.space == p.space
   psolve(r, out=s) # line 432
   am = s.dot(r) # line 433
   s.copy(out=p) # line 434
   ...
   # This line asserts isinstance(p, A.domain)
   v  = A.dot(p, out=v) # line 453
```
Combining all, we have v.space == r.space == s.space == p.space, but v in A.codomain, p in A.domain.

This means that for now, domain and codomain must be the same. Nonetheless I added a test (commented out) that should work once (or if ever) we find a way to deal with this issue.

Reply:
a5823654f3eeb2d09af85ab75e2ae202f14b1c04